File: app/src/main/java/com/knotworking/numbers/counter/CounterDiffCallback.java
package com.knotworking.numbers.counter;

import android.support.v7.util.DiffUtil;

import java.util.List;

/**
 * Created by BRL on 22/04/17.
 */

public class CounterDiffCallback extends DiffUtil.Callback {

    private List<CounterItem> oldList;
    private List<CounterItem> newList;

    public CounterDiffCallback(List<CounterItem> oldList, List<CounterItem> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList != null ? oldList.size() : 0;
    }

    @Override
    public int getNewListSize() {
        return newList != null ? newList.size() : 0;
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return oldList.get(oldItemPosition).getId() == newList.get(newItemPosition).getId();
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        CounterItem oldItem = oldList.get(oldItemPosition);
        CounterItem newItem = newList.get(newItemPosition);
        return oldItem.equals(newItem);
    }
}

File: app/src/main/java/com/knotworking/numbers/counter/CounterDiffCallback.kt
package com.knotworking.numbers.counter

import android.support.v7.util.DiffUtil

/**
 * Created by BRL on 22/04/17.
 */

class CounterDiffCallback(private val oldList: List<CounterItem>, private val newList: List<CounterItem>) : DiffUtil.Callback() {

    override fun getOldListSize(): Int {
        return oldList.size
    }

    override fun getNewListSize(): Int {
        return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldList[oldItemPosition].id == newList[newItemPosition].id
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldItem = oldList[oldItemPosition]
        val newItem = newList[newItemPosition]
        return oldItem == newItem
    }
}
